Std/Tactic/TryThis.lean

```diff
@@ -1,42 +1,144 @@
+      changes: { [props.pos.uri]: [{ range: props.range, newText: props.suggestion }] }
+    })
+  }
+  return e('div', {className: 'ml1'}, e('pre', {className: 'font-code pre-wrap'}, [
+    'Try this: ',
+    e('a', {onClick, className: 'link pointer dim', title: 'Apply suggestion'}, props.suggestion),
+    props.info
+  ]))
+}"
+
+/--
+This is a code action provider that looks for `TryThisInfo` nodes and supplies a code action to
+apply the replacement.
+-/
+@[codeActionProvider] def tryThisProvider : CodeActionProvider := fun params snap => do
+  let doc ← readDoc
+  pure <| snap.infoTree.foldInfo (init := #[]) fun _ctx info result => Id.run do
+    let .ofUserWidgetInfo { stx, widgetId := ``tryThisWidget, props } := info | result
+    let some stxRange := stx.getRange? | result
+    let stxRange := doc.meta.text.utf8RangeToLspRange stxRange
+    unless stxRange.start.line ≤ params.range.end.line do return result
+    unless params.range.start.line ≤ stxRange.end.line do return result
+    let .ok newText := props.getObjValAs? String "suggestion" | panic! "bad type"
+    let .ok range := props.getObjValAs? Lsp.Range "range" | panic! "bad type"
+    result.push {
+      eager.title := "Apply 'Try this'"
+      eager.kind? := "refactor"
+      eager.edit? := some <| .ofTextEdit params.textDocument.uri { range, newText }
+    }
 
 /-- Replace subexpressions like `?m.1234` with `?_` so it can be copy-pasted. -/
 partial def replaceMVarsByUnderscores [Monad m] [MonadQuotation m]
     (s : Syntax) : m Syntax :=
-  s.replaceM fun s => if let `(?$_:ident) := s then `(?_) else pure none
+  s.replaceM fun s => do
+    let `(?$id:ident) := s | pure none
+    if id.getId.hasNum || id.getId.isInternal then `(?_) else pure none
 
 /-- Delaborate `e` into an expression suitable for use in `refine`. -/
 def delabToRefinableSyntax (e : Expr) : TermElabM Term :=
   return ⟨← replaceMVarsByUnderscores (← delab e)⟩
 
-/-- Add a "try this" suggestion. (TODO: this depends on code action support) -/
-def addSuggestion [Monad m] [MonadLog m] [AddMessageContext m] [MonadOptions m]
-    (origStx : Syntax) (suggestion : Syntax) : m Unit :=
-  logInfoAt origStx m!"{suggestion}"
+/-- Add a "try this" suggestion. This has three effects:
+
+* An info diagnostic is displayed saying `Try this: <suggestion>`
+* A widget is registered, saying `Try this: <suggestion>` with a link on `<suggestion>` to apply
+  the suggestion
+* A code action `Apply 'Try this'` is added, which will apply the suggestion.
 
-/-- Add a `exact e` or `refine e` suggestion. (TODO: this depends on code action support) -/
-def addExactSuggestion (origTac : Syntax) (e : Expr) : TermElabM Unit := do
+The parameters are:
+* `ref`: the span of the info diagnostic
+* `suggestion`: the replacement syntax
+* `suggestionForMessage?`: the message to display in the info diagnostic (only).
+  The widget message uses only `suggestion`. If not provided, it uses `suggestion` in both places.
```

Comment:
```suggestion
  The widget message uses only `suggestion`. If not provided, `suggestion` is used in both places.
```
Otherwise it sounds like "the widget message uses `suggestion` in both places" which is presumably not what was meant.